src/main/java/graphql/analysis/QueryTraversal.java

```diff
@@ -146,79 +140,67 @@ public Object visitInlineFragment(InlineFragment inlineFragment, TraverserContex
+            context.setVar(QueryTraversalContext.class, new QueryTraversalContext(fragmentCondition, parentEnv.getEnvironment()));
+            return TraversalControl.CONTINUE;
         }
 
         @Override
-        public Object visitFragmentSpread(FragmentSpread fragmentSpread, TraverserContext<Selection> context) {
+        public TraversalControl visitFragmentSpread(FragmentSpread fragmentSpread, TraverserContext<Node> context) {
+            if (context.getVar(LeaveOrEnter.class) == LEAVE) {
+                return TraversalControl.CONTINUE;
+            }
             if (!conditionalNodes.shouldInclude(variables, fragmentSpread.getDirectives()))
-                return TraverserMarkers.ABORT;
+                return TraversalControl.ABORT;
 
             FragmentDefinition fragmentDefinition = fragmentsByName.get(fragmentSpread.getName());
             if (!conditionalNodes.shouldInclude(variables, fragmentDefinition.getDirectives()))
-                return TraverserMarkers.ABORT;
+                return TraversalControl.ABORT;
 
             QueryTraversalContext parentEnv = context
                     .parentContext()
                     .getVar(QueryTraversalContext.class);
 
             GraphQLCompositeType typeCondition = (GraphQLCompositeType) schema.getType(fragmentDefinition.getTypeCondition().getName());
 
-            return context
-                .setVar(QueryTraversalContext.class, new QueryTraversalContext(typeCondition, parentEnv.getEnvironment()));
+            context
+                    .setVar(QueryTraversalContext.class, new QueryTraversalContext(typeCondition, parentEnv.getEnvironment()));
+            return TraversalControl.CONTINUE;
         }
 
         @Override
-        public Object visitField(Field field, TraverserContext<Selection> context) {
-            if (!conditionalNodes.shouldInclude(variables, field.getDirectives()))
-                return TraverserMarkers.ABORT; // stop recursion
-
+        public TraversalControl visitField(Field field, TraverserContext<Node> context) {
             QueryTraversalContext parentEnv = context
                     .parentContext()
                     .getVar(QueryTraversalContext.class);
 
             GraphQLFieldDefinition fieldDefinition = Introspection.getFieldDef(schema, parentEnv.getType(), field.getName());
             Map<String, Object> argumentValues = valuesResolver.getArgumentValues(schema.getFieldVisibility(), fieldDefinition.getArguments(), field.getArguments(), variables);
+            QueryVisitorEnvironment environment = new QueryVisitorEnvironment(field, fieldDefinition, parentEnv.getType(), parentEnv.getEnvironment(), argumentValues);
+
+            LeaveOrEnter leaveOrEnter = context.getVar(LeaveOrEnter.class);
+            if (leaveOrEnter == LEAVE) {
+                if (!preOrder) {
+                    visitFieldCallback.visitField(environment);
```

Comment:
separating callback `Consumers` into preOrder and postOrder Consumer in my mind made the core a little more readable, in a sense that there were no redundant check like this `if (!preOrder)`.

Additionally, there is redundant code above the line 181 that is executed twice. Is that intentional?

Reply:
agreed. changed it. But not sure what redundant code you mean ... I don't see it.